HARK/core.py

```diff
@@ -18,11 +19,201 @@
+            self._age_inv.append(key)
+
+    def __add_to_time_vary(self, key):
+        """
+        Adds parameter name to varying list and removes from invariant list.
+
+        Parameters
+        ----------
+        key : str
+            parameter name
+        """
+        if key in self._age_inv:
+            self._age_inv.remove(key)
+        if key not in self._age_var:
+            self._age_var.append(key)
+
+    def __getitem__(self, age_or_key):
+        """
+        If age_or_key is an integer, returns a Parameters object with the parameters
+        that apply to that age. This includes all invariant parameters and the
+        `age_or_key`th element of all age-varying parameters. If age_or_key is a string,
+        it returns the value of the parameter with that name.
+
+        Parameters
+        ----------
+        age_or_key : int or str
+            Age or key of parameter(s)
+
+
+        Returns
+        -------
+        Parameters or value
+            Parameters object with parameters that apply to age `age_or_key` or value of
+            parameter with name `age_or_key`.
+        """
+        if isinstance(age_or_key, int):
+            # return parameters at age
+            assert age_or_key < self._term_age
+
+            params = {}
+            for key in self._age_inv:
+                params[key] = getattr(self, key)
+            for key in self._age_var:
+                params[key] = getattr(self, key)[age_or_key]
+            return Parameters(**params)
+
+        elif isinstance(age_or_key, str):
+            return getattr(self, age_or_key)
+
+    def __setitem__(self, key, value):
+        """
+        Sets the value of a parameter.
+
+        Parameters
+        ----------
+        key : str
+            name of parameter
+        value : Any
+            value of parameter
+
+        """
+        if not isinstance(key, str):
+            raise ValueError("Parameters must be set with a string key")
+        self.__setattr__(key, value)
+
+    def __setattr__(self, key, value):
+        if key.startswith("_"):
+            # Handle setting internal attributes normally
+            super().__setattr__(key, value)
+        else:
+            # Handle setting parameters as dictionary items
+            new_value = self.__infer_dims__(key, value)
+            super().__setattr__(key, new_value)
+
+    def keys(self):
+        return self._age_inv + self._age_var
+
+    def values(self):
+        return [getattr(self, key) for key in self.keys()]
+
+    def items(self):
+        return [(key, getattr(self, key)) for key in self.keys()]
+
+    def __iter__(self):
+        return iter(self.keys())
+
+    def __deepcopy__(self, memo):
+        return Parameters(**deepcopy(self.to_dict(), memo))
+
+    def to_dict(self):
+        return {key: getattr(self, key) for key in self.keys()}
+
+    def to_namedtuple(self):
+        return namedtuple("Parameters", self.keys())(**self.to_dict())
```

Comment:
What about a `from_namedtuple`? See above about time slicing a Parameters object.

Reply:
I don't know that we need a special case for `from_namedtuple`. With `namedtuple `you can do `namedtuple._asdict()`, so to create a Parameters object from a `namedtuple `you'd just have to do `Parameters(**namedtuple._asdict()) `